flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,321 @@
+    parameters.add(UseQuadTreeParam, UseQuadTree)
+    this
+  }
+
+
+}
+
+object KNN {
+
+  case object K extends Parameter[Int] {
+    val defaultValue: Option[Int] = Some(5)
+  }
+
+  case object DistanceMetric extends Parameter[DistanceMetric] {
+    val defaultValue: Option[DistanceMetric] = Some(EuclideanDistanceMetric())
+  }
+
+  case object Blocks extends Parameter[Int] {
+    val defaultValue: Option[Int] = None
+  }
+
+  case object UseQuadTreeParam extends Parameter[Boolean] {
+    val defaultValue: Option[Boolean] = None
+  }
+
+
+  def apply(): KNN = {
+    new KNN()
+  }
+
+  /** [[FitOperation]] which trains a KNN based on the given training data set.
+    * @tparam T Subtype of [[org.apache.flink.ml.math.Vector]]
+    */
+
+  implicit def fitKNN[T <: FlinkVector : TypeInformation] = new FitOperation[KNN, T] {
+    override def fit(
+                      instance: KNN,
+                      fitParameters: ParameterMap,
+                      input: DataSet[T]): Unit = {
+      val resultParameters = instance.parameters ++ fitParameters
+
+      require(resultParameters.get(K).isDefined, "K is needed for calculation")
+
+      val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+      val partitioner = FlinkMLTools.ModuloKeyPartitioner
+      val inputAsVector = input.asInstanceOf[DataSet[FlinkVector]]
+
+      instance.trainingSet = Some(FlinkMLTools.block(inputAsVector, blocks, Some(partitioner)))
+    }
+  }
+
+  /** [[PredictDataSetOperation]] which calculates k-nearest neighbors of the given testing data
+    * set.
+    * @tparam T Subtype of [[Vector]]
+    * @return The given testing data set with k-nearest neighbors
+    */
+
+  implicit def predictValues[T <: FlinkVector : ClassTag : TypeInformation] = {
+    new PredictDataSetOperation[KNN, T, (FlinkVector, Array[FlinkVector])] {
+      override def predictDataSet(
+                                   instance: KNN,
+                                   predictParameters: ParameterMap,
+                                   input: DataSet[T]):
+                                   DataSet[(FlinkVector, Array[FlinkVector])] = {
+        val resultParameters = instance.parameters ++ predictParameters
+
+        instance.trainingSet match {
+          case Some(trainingSet) =>
+            val k = resultParameters.get(K).get
+            val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+            val metric = resultParameters.get(DistanceMetric).get
+            val partitioner = FlinkMLTools.ModuloKeyPartitioner
+
+            // attach unique id for each data
+            val inputWithId: DataSet[(Long, T)] = input.zipWithUniqueId
+
+            // split data into multiple blocks
+            val inputSplit = FlinkMLTools.block(inputWithId, blocks, Some(partitioner))
+
+            // join input and training set
+            val crossed = trainingSet.cross(inputSplit).mapPartition {
```

Comment:
It's true. To do it automatically it's tricky. What one could do is either counting both data sources and then decide on the number of elements. Or one could implement a way for the user to specify which side is larger. Like it's done for the join operator with `joinWithTiny` and `joinWithLarge`.

I actually only commented on it because `cross` can be a really costly operation because both sides of the operator have to be effectively shuffled.